sim-lib/src/lib.rs

```diff
@@ -994,29 +1077,46 @@ async fn produce_simulation_results(
     nodes: HashMap<PublicKey, Arc<Mutex<dyn LightningNode + Send>>>,
     mut output_receiver: Receiver<SimulationOutput>,
     results: Sender<(Payment, PaymentResult)>,
-    shutdown: Listener,
+    listener: Listener,
+    shutdown: Trigger,
 ) {
     log::debug!("Simulation results producer started.");
 
     let mut set = tokio::task::JoinSet::new();
 
     loop {
-        tokio::select! {
+        select! {
             biased;
-            _ = shutdown.clone() => break,
+            _ = listener.clone() =>
+                break
+            ,
             output = output_receiver.recv() => {
                 match output{
                     Some(simulation_output) => {
                         match simulation_output{
                             SimulationOutput::SendPaymentSuccess(payment) => {
-                                let source_node = nodes.get(&payment.source).unwrap().clone();
-                                set.spawn(track_payment_result(
-                                    source_node, results.clone(), payment, shutdown.clone(),
-                                ));
+                                if let Some(source_node) = nodes.get(&payment.source) {
```

Comment:
If we're getting rid of the `unwrap` here, we should error out if we can' t find the source node, right? Because we've generated this successful payment from a node that we don't have in our node set.

Reply:
Updated to handle this possibility.